### ISSUE TITLE
Troubleshooting GetEstimatedRemainingWorkPerPartitionAsync

### ISSUE THREAD

**Issue by bartelink** [Azure/azure-documentdb-changefeedprocessor-dotnet#126]
I'm periodically using `GetEstimatedRemainingWorkPerPartitionAsync` to establish the lag my lease has (especially relevant to #123) versus the present write position.

I have read access to a closed source implementation of code that successfully uses this correctly, and it both fits the bill and seems to work well.

Problem: Using very similar code (I'll add a link in due course, but suffice to say the implementation is pretty boring (which absolutely does not rule out a very simple case of PEBKAC)), I'm getting an empty array of responses, and not seeing any clues as to the reasons why this might be failing in the LibLog emissions.

Question: (which I can and will solve by poring over the source if push comes to shove) how can this happen - what things need to be in place ? is it possible that some accounts do not have the changefeed correctly configured ? Is it reasonable to  expect either an error message in the log or >=1 entry in the response array per time I call ?

Any advice wrt what to expect in terms of failures cases wrt this API would be appreciated.

**Comment by bartelink** [comment]
Can you double confirm that there's no _modifier_ switches that might alter the behavior I'd see on a given CosmosDb account? (The changefeed is working in all other respects; I'm not expecting it to be that high level a thing as an changefeed on/off switch - I'm just trying to rule out the potential that it's working in the other account due to some privilege or mode bestowed on that specific account by the setting of a specific flag and/or that this account is missing a modifier)

In [my code](https://github.com/jet/equinox/blob/058ecd4c961a81dce610ed8161f61c9a0c3341f5/src/Equinox.Cosmos.Projection/Projection.fs#L133), the call always returns `[]`.

From looking in the impl, I see the body of the method can do that in a given set of states.

Rewording the question:
a) can you see anything obvious wrong with my code which might lead to this? I copied it from somewhere else with minimal changes, and I know it works in that context
b) OR can you explain (based on your deeper knowledge of how the overall codebase works), the potential things that I might be doing wrong and/or assumptions implicit in how it's supposed to be used which would lead this to happen

The quickstart in the readme in my codebase is effectively a set of Steps To Reproduce, but I'm thinking I'm making some assumption  that can likely be flushed out by you explaining Dos and Dont's as to how to ensure he estimator works.

Wild guess: I'm always supplying a `leasePrefix`, could there be a bug due to that?

**Comment by ealsur** [comment]
@bartelink What's the status of the Leases collection when you call the estimator in your code? Was it already initialized? Does it have Leases with continuations (ContinuationToken) already set?
I'm not proficient in F# in order to debug if there are any errors or code paths that might lead to a different behavior.

**Comment by bartelink** [comment]
I run the estimator in a timed loop alongside the projector - I've definitely run it with and without there being leases in there and never seen it produce results. Let me dig in to verify for sure (might be a few days though). Wrt it being F#, you can just set a breakpoint ;P - the callback definitely gets triggered (but with an empty array) and there's no reason the language should have any effect (it is as if I got the name of the projection wrong though - my first thing to check will be whether that leg of the builder honors the leasePrefix or not - if the BuildEstimator did not honor it, that would explain it).

**Comment by ealsur** [comment]
I would check what's the status of the leases when the Estimator run. For example:
1. Let the Processor run a couple of times and stop it
2. Make some changes in the collection
3. Run the Estimator and check results

The Estimator checks the current state of leases vs the current state of the Feed. If there are no leases, or the leases have no ContinuationToken (they have not been checkpointed ever), then the comparison would yield an empty result, because there is no beginning time to compare against.

https://github.com/Azure/azure-documentdb-changefeedprocessor-dotnet/blob/97cd308b1f9ec85cae8f6b8740b0a21dbb5524ee/src/DocumentDB.ChangeFeedProcessor/PartitionManagement/RemainingWorkEstimator.cs#L54

**Comment by bartelink** [comment]
Makes sense. If it's alright, I'll leave this issue open and report back what the problem was when I figure it out - I won't be long (please close the issue should I unexpectedly go dark for >1 week!)

**Comment by bartelink** [comment]
Don't have a full repro, but have observed that removing the `LeaseId` (I have a wrapper that makes it mandatory to supply one) makes it work. This implies there is some weakness or inconsistency in how the progress is derived in that way.

I observed that establishing a single lease without a prefix results in the function starting to returning values (but the values never move)

Example output (using a LeaseId where the `aux` collection has previously ran without a leaseId - note a) that no progress is registered for > 1h despite operating with auto-acks b) no data is returned at all until I run once without a leaseId):
![image](https://user-images.githubusercontent.com/206668/52382868-d03df000-2a6e-11e9-83d8-a731cd53d5c3.png)
